### PR TITLE
Windows support

### PR DESCRIPTION
**Pull request by nkaretnikov**
See #507. Original PR: #559.

### CODE REVIEW

**Comment by nkaretnikov** [comment]
Removed some original CI changes while rebasing, since main uses miniforge.

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -51,7 +52,7 @@ jobs:
 
       - name: "Unit tests ✅"
         run: |
-          pytest tests
+          pytest -vvv tests
```
Comment (nkaretnikov) [inline .github/workflows/tests.yaml:55]: Remove.

File: conda-store-server/conda_store_server/server/app.py [hunk 2]
```diff
@@ -335,6 +336,10 @@ def start(self):
 
         # start worker if in standalone mode
         if self.standalone:
+            address = "localhost" if self.address == "0.0.0.0" else self.address
+            print(
+                f"Starting standalone conda-store server at http://{address}:{self.port}"
+            )
```
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/server/app.py:342]: Remove.

File: conda-store-server/conda_store_server/utils.py [hunk 3]
```diff
@@ -50,13 +50,52 @@ def chdir(directory: pathlib.Path):
...
+            if st.st_ino in have:
+                continue
+            have.add(st.st_ino)
+            apparent_total_bytes += st.st_size
+        for d in dirnames:
+            dp = os.path.join(dirpath, d)
+            if os.path.islink(dp):
+                apparent_total_bytes += os.lstat(dp).st_size
+
+    return apparent_total_bytes
```
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/utils.py:83]: Needs clean up.

File: conda-store-server/conda_store_server/utils.py [hunk 4]
```diff
@@ -50,13 +50,52 @@ def chdir(directory: pathlib.Path):
...
         cmd = ["du", "-sb", str(path)]
+    else:
+        return str(du(path))
 
-    return subprocess.check_output(cmd, encoding="utf-8").split()[0]
+    output = subprocess.check_output(cmd, encoding="utf-8").split()[0]
+    if sys.platform == "darwin":
+        # mac du does not have the -b option to return bytes
+        output = str(int(output) * 512)
+    return output
```
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/utils.py:98]: Move macOS code to a separate PR.

File: conda-store-server/environment-windows-dev.yaml [hunk 5]
```diff
@@ -0,0 +1,56 @@
+name: conda-store-server-dev
+channels:
+  - conda-forge
+  - Microsoft
+dependencies:
```
Comment (nkaretnikov) [inline conda-store-server/environment-windows-dev.yaml:5]: Needs to match other dev envs on main.

File: conda-store-server/environment-windows-dev.yaml [hunk 6]
```diff
@@ -0,0 +1,56 @@
+name: conda-store-server-dev
```
Comment (nkaretnikov) [inline conda-store-server/environment-windows-dev.yaml:1]: Needs to match other dev envs on main.

File: .github/workflows/tests.yaml [hunk 7]
```diff
@@ -19,13 +19,14 @@ jobs:
...
-        # cannot run on windows due to needing fake-chroot for conda-docker
-        os: ["ubuntu", "macos"]
+        os: ["ubuntu", "macos", "windows"]
         include:
           - os: ubuntu
             environment-file: conda-store-server/environment-dev.yaml
           - os: macos
             environment-file: conda-store-server/environment-macos-dev.yaml
+          - os: windows
+            environment-file: conda-store-server/environment-windows-dev.yaml
```
Comment (nkaretnikov) [inline .github/workflows/tests.yaml:29]: Check that this works. Might need additional changes.

File: conda-store-server/tests/test_usage.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    # on Mac and Linux but uses the Python du() function on Windows.
+    val = disk_usage(test_file)
+    assert isinstance(val, str)
+    assert 1000 <= int(val) <= 1500
+    assert 1000 <= du(test_file) <= 1500
+
+    val = disk_usage(test_dir)
+    assert isinstance(val, str)
+    assert 2000 + dir_size <= int(val) <= 2700 + dir_size
+    assert 2000 + dir_size <= du(test_dir) <= 2700 + dir_size
```
Comment (nkaretnikov) [inline conda-store-server/tests/test_usage.py:32]: This could use more tests. Bounds are 50% of file size, too.

File: conda-store-server/tests/test_actions.py [hunk 9]
```diff
@@ -155,6 +164,7 @@ def test_remove_conda_prefix(tmp_path, simple_conda_lock):
     assert not conda_prefix.exists()
 
 
+@pytest.mark.skipif(sys.platform == "win32", reason="permissions are not supported on Windows")
```
Comment (nkaretnikov) [inline conda-store-server/tests/test_actions.py:167]: Use the same ON_WIN check here.

**Comment by nkaretnikov** [comment]
Note: Windows test takes 25 min. macOS: 14 min. Ubuntu: 10 min.

File: conda-store-server/conda_store_server/utils.py (hunk 4 above)
Comment (nkaretnikov) [reply conda-store-server/conda_store_server/utils.py:90]: This is needed for the du test to pass on macOS CI runner.

File: conda-store-server/tests/test_actions.py (hunk 9 above)
Comment (nkaretnikov) [reply conda-store-server/tests/test_actions.py:167]: I'll do the other way around and use the win32 check instead of the constant everywhere.

File: conda-store-server/conda_store_server/app.py [hunk 10]
```diff
@@ -302,21 +304,24 @@ def _default_celery_results_backend(self):
     )
 
     default_uid = Integer(
-        os.getuid(),
+        None if ON_WIN else os.getuid(),
```
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/app.py:307]: Use sys.platform here instead of the constant.

File: conda-store-server/conda_store_server/schema.py [hunk 11]
```diff
@@ -194,20 +196,20 @@ class Settings(BaseModel):
         metadata={"global": True},
     )
 
-    default_uid: int = Field(
-        os.getuid(),
+    default_uid: int | None = Field(
+        None if ON_WIN else os.getuid(),
```
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/schema.py:200]: Use sys.platform here instead of the constant.

File: conda-store-server/conda_store_server/schema.py [hunk 12]
```diff
@@ -194,20 +196,20 @@ class Settings(BaseModel):
         metadata={"global": True},
     )
 
-    default_uid: int = Field(
-        os.getuid(),
+    default_uid: int | None = Field(
```
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/schema.py:199]: Use Optional instead of `| None`.

File: conda-store-server/tests/test_usage.py (hunk 8 above)
Comment (nkaretnikov) [reply conda-store-server/tests/test_usage.py:33]: Bounds:

```
ubuntu

2023-10-22T18:43:40.3618723Z tests/test_usage.py XXXXXXXXXXXXXX: 1000 1000
2023-10-22T18:43:40.3619644Z XXXXXXXXXXXXXX: 4096 6096 6096

macos

2023-10-22T18:49:11.6388490Z tests/test_usage.py XXXXXXXXXXXXXX: 1024 1000
2023-10-22T18:49:11.6389010Z XXXXXXXXXXXXXX: 64 2560 2160

windows

2023-10-22T18:56:13.8420208Z tests\test_usage.py XXXXXXXXXXXXXX: 1000 1000
2023-10-22T18:56:13.8420808Z XXXXXXXXXXXXXX: 0 2000 2000


```

Comment (nkaretnikov) [reply conda-store-server/tests/test_usage.py:33]: I'll keep this as is right now, but it might be worth it to revise later. Especially, on macOS.

File: conda-store-server/conda_store_server/action/base.py [hunk 13]
```diff
@@ -18,7 +18,7 @@ def wrapper(*args, **kwargs):
             # redirect stdout -> action_context.stdout
             stack.enter_context(contextlib.redirect_stdout(action_context.stdout))
 
-            # redirect stderr -> action_context.stdout
+            # redirect stderr -> action_context.stderr
```
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/action/base.py:21]: Comment doesn't match code. Revert this.

File: conda-store-server/tests/test_actions.py [hunk 14]
```diff
@@ -15,9 +15,17 @@ def test_action_decorator():
...
+            # echo is not a separate program on Windows
+            context.run(["cmd", "/c", "echo subprocess"])
+            context.run("echo subprocess_stdout", shell=True)
+            context.run("echo subprocess_stderr>&2", shell=True)
+            context.run("echo subprocess_stderr_no_redirect>&2", shell=True, redirect_stderr=False)
+        else:
+            context.run(["echo", "subprocess"])
+            context.run("echo subprocess_stdout", shell=True)
+            context.run("echo subprocess_stderr 1>&2", shell=True)
+            context.run("echo subprocess_stderr_no_redirect 1>&2", shell=True, redirect_stderr=False)
```
Comment (nkaretnikov) [inline conda-store-server/tests/test_actions.py:28]: Only the first command needs to be inside if-else.

File: conda-store-server/tests/test_usage.py [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+    dir_size = du(test_dir)
+    assert abs(dir_size - int(disk_usage(test_dir))) <= 1000
+
+    test_file = test_dir / "test_file"
+    test_file.write_text("a"*1000)
+    test_file2 = test_dir / "test_file2"
+    test_file2.write_text("b"*1000)
+    # Test hard links
+    test_file_hardlink = test_dir / "test_file_hardlink"
+    test_file_hardlink.hardlink_to(test_file)
```
Comment (nkaretnikov) [inline conda-store-server/tests/test_usage.py:19]: Note for reviewers: this is tested where test_dir size is tested as a whole.

File: conda-store-server/conda_store_server/action/base.py [hunk 16]
```diff
@@ -38,20 +38,23 @@ class ActionContext:
...
             **kwargs,
             stdout=subprocess.PIPE,
-            stderr=subprocess.STDOUT,
+            stderr=subprocess.STDOUT if redirect_stderr else subprocess.PIPE,
             encoding="utf-8",
         )
         self.stdout.write(result.stdout)
+        if not redirect_stderr:
+            self.stderr.write(result.stderr)
         return result
```
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/action/base.py:60]: See https://github.com/conda-incubator/conda-store/pull/559#issuecomment-1756298903 for context on redirects.

File: conda-store-server/conda_store_server/server/auth.py [hunk 17]
```diff
@@ -436,7 +436,9 @@ async def post_login_method(
             samesite="strict",
             domain=self.cookie_domain,
             # set cookie to expire at same time as jwt
-            max_age=(authentication_token.exp - datetime.datetime.utcnow()).seconds,
+            max_age=int(
+                (authentication_token.exp - datetime.datetime.utcnow()).total_seconds()
+            ),
```
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/server/auth.py:441]: See https://github.com/conda-incubator/conda-store/pull/559/files#r1364084772 for context on this.

File: conda-store-server/tests/test_actions.py (hunk 14 above)
Comment (nkaretnikov) [reply conda-store-server/tests/test_actions.py:28]: Nope, this is parsed slightly differently on Windows. Whitespace would be included into the output with redirect_stderr. So this is fine.

File: docs/user_guide.md [hunk 18]
```diff
@@ -73,6 +73,10 @@ conda-unpack
 
 ### Docker Registry
 
+```{note}
```
Comment (nkaretnikov) [inline docs/user_guide.md:76]: Not sure if {note} syntax is supported.

**Comment by nkaretnikov** [comment]
Manually tested this env via the admin UI:

```
channels:
- conda-forge
dependencies:
- python
- pip:
  - nothing
- ipykernel
- pytest
- requests
description: ''
name: test-env
prefix: null
variables: null
```

Building, rebuilding, and deletion work.

**Comment by nkaretnikov** [comment]
@costrouc This is Aaron's Windows PR with some minor changes from me. PTAL and merge if OK.

**Comment by jaimergp** [comment]
> the post set up env on Windows takes 3 mins

You can do `run-post: false` (or similar name) on setup-miniconda. It's an ephemeral VM after all.

File: conda-store-server/conda_store_server/orm.py [hunk 19]
```diff
@@ -513,10 +513,11 @@ def update_packages(self, db, subdirs=None):
                 package_builds[package_key].append(new_package_build_dict)
             logger.info("CondaPackageBuild objects created")
 
-            batch_size = 1000
+            # sqlite3 has a max expression depth of 1000
+            batch_size = 990
```
Comment (costrouc) [inline conda-store-server/conda_store_server/orm.py:517]: I bet there is an interesting story here? Is this only something that showed up on Windows?

File: conda-store-server/conda_store_server/app.py [hunk 20]
```diff
@@ -302,21 +302,24 @@ def _default_celery_results_backend(self):
     )
 
     default_uid = Integer(
-        os.getuid(),
+        None if sys.platform == "win32" else os.getuid(),
```
Comment (costrouc) [inline conda-store-server/conda_store_server/app.py:305]: I see `sys.platform == "win32"` in many places. Should we have a constant that we create for this? Since much of our code only has a special path for Windows?

File: conda-store-server/conda_store_server/server/app.py [hunk 21]
```diff
@@ -198,9 +199,9 @@ def trim_slash(url):
         app = FastAPI(
             title="conda-store",
             version=__version__,
-            openapi_url=os.path.join(self.url_prefix, "openapi.json"),
-            docs_url=os.path.join(self.url_prefix, "docs"),
-            redoc_url=os.path.join(self.url_prefix, "redoc"),
+            openapi_url=posixpath.join(self.url_prefix, "openapi.json"),
```
Comment (costrouc) [inline conda-store-server/conda_store_server/server/app.py:202]: I'll admit using `os.path...` for urls is a hack and I should have used something like yarl for example. Thoughts? Since url paths are not the same as file paths? Maybe posixpath is the right thing to use here. It certainly seems better than os.path.

File: conda-store-server/conda_store_server/server/auth.py (hunk 17 above)
Comment (costrouc) [reply conda-store-server/conda_store_server/server/auth.py:441]: A fun bug :smile: glad we got to the bottom of this one.

File: conda-store-server/conda_store_server/utils.py [hunk 22]
```diff
@@ -50,13 +50,44 @@ def chdir(directory: pathlib.Path):
...
+
+
 def disk_usage(path: pathlib.Path):
     if sys.platform == "darwin":
         cmd = ["du", "-sAB1", str(path)]
-    else:
+    elif sys.platform == "linux":
         cmd = ["du", "-sb", str(path)]
+    else:
+        return str(du(path))
```
Comment (costrouc) [inline conda-store-server/conda_store_server/utils.py:84]: Good to see the fallback is this approach :+1: If they are using cygwin or https://learn.microsoft.com/en-us/sysinternals/downloads/du it may still be possible they have du.

File: conda-store-server/conda_store_server/utils.py [hunk 23]
```diff
@@ -50,13 +50,44 @@ def chdir(directory: pathlib.Path):
...
 
-    return subprocess.check_output(cmd, encoding="utf-8").split()[0]
+    output = subprocess.check_output(cmd, encoding="utf-8").split()[0]
+    if sys.platform == "darwin":
+        # mac du does not have the -b option to return bytes
+        output = str(int(output) * 512)
```
Comment (costrouc) [inline conda-store-server/conda_store_server/utils.py:89]: Why 512? This is a weird size since it is a half a kb.

File: conda-store-server/conda_store_server/worker/app.py [hunk 24]
```diff
@@ -71,9 +71,18 @@ def start(self):
         argv = [
             "worker",
             "--loglevel=INFO",
-            "--beat",
         ]
 
+        # The default Celery pool requires this on Windows. See
+        # https://stackoverflow.com/questions/37255548/how-to-run-celery-on-windows
```
Comment (costrouc) [inline conda-store-server/conda_store_server/worker/app.py:77]: Great glad this fix was minimal.

File: conda-store-server/tests/test_usage.py (hunk 8 above)
Comment (costrouc) [reply conda-store-server/tests/test_usage.py:32]: > The exact values depend on the block size and other details

Oh yeah this is a fun detail :smile: guess you can't be smaller than the blocksize and also can only take up an integer number of block sizes.

**Review by costrouc — APPROVED** [review]
Overall these changes do not look controversial and I'm surprised that the windows fixes were actually smaller than I expected. I was really expecting issues with pathlib etc but it looks like most of those bugs didn't exist.

File: conda-store-server/conda_store_server/app.py (hunk 20 above)
Comment (costrouc) [reply conda-store-server/conda_store_server/app.py:305]: More I think about it I don't think this is necissary

File: conda-store-server/conda_store_server/orm.py (hunk 19 above)
Comment (nkaretnikov) [reply conda-store-server/conda_store_server/orm.py:517]: I haven't seen this personally, but it's mentioned in one of the Aaron's commits:

```
Lower the batch size for updating packages from 1000 to 990
1000 hits the sqlite3 max expression depth and results in an error.
```

Seems harmless, so I decided to keep it.

File: conda-store-server/conda_store_server/app.py (hunk 20 above)
Comment (nkaretnikov) [reply conda-store-server/conda_store_server/app.py:305]: There was a constant in Aaron's original code. Well, two constants, in different files. Since you need to share this between multiple files, you'd have to put it in `__init__` to avoid import issues. But we don't do this for any other platforms, so I decided to simply use `sys.platform`, as everywhere else, to avoid introducing new entities.

File: conda-store-server/conda_store_server/server/app.py (hunk 21 above)
Comment (nkaretnikov) [reply conda-store-server/conda_store_server/server/app.py:202]: A separate entity for URLs would be better, yes. But I'm not going to block this PR just because of this issue, since everything seems to work fine.

File: conda-store-server/conda_store_server/utils.py (hunk 22 above)
Comment (nkaretnikov) [reply conda-store-server/conda_store_server/utils.py:84]: We've already looked at du from sysinternals. Besides being a separate tool, it also produces slightly different results. And the output format is also different. As for cygwin, the context for this work is to be able to use native Windows tooling.

File: conda-store-server/conda_store_server/utils.py (hunk 23 above)
Comment (nkaretnikov) [reply conda-store-server/conda_store_server/utils.py:89]: A comment from a discussion on the original PR: "Note that the numbers will be off a little bit (by less than 512) on Mac because du rounds up to the nearest multiple of 512 when converting bytes to blocks."

I'm still not certain that macOS code is correct here, but it does pass this new test (where we check all three platforms), so I'm going to keep this and will take a look later.

File: conda-store-server/conda_store_server/worker/app.py (hunk 24 above)
Comment (nkaretnikov) [reply conda-store-server/conda_store_server/worker/app.py:77]: It's still not clear whether it means that Celery works as intended, since it's not officially supported on Windows. But it does run, so we'll just use this for now.
